File: typescript/src/cache.ts
import type { FeedId, PriceUpdate } from "./types.js";
import type { PriceTransport } from "./transport.js";

interface Entry {
  update: PriceUpdate;
  fetchedAtMs: number;
}

/**
 * A TTL cache decorator that wraps any {@link PriceTransport}.
 *
 * `CachingTransport` *is a* `PriceTransport`, so it composes with the generic
 * client exactly like any other backend. Entries older than `ttlMs` are
 * refreshed on access.
 */
export class CachingTransport implements PriceTransport {
  private readonly cache = new Map<string, Entry>();

  constructor(
    private readonly inner: PriceTransport,
    private readonly ttlMs: number,
  ) {}

  async fetch(feed: FeedId): Promise<PriceUpdate> {
    const key = feed.asString();
    const hit = this.cache.get(key);
    if (hit && Date.now() - hit.fetchedAtMs < this.ttlMs) {
      return hit.update;
    }
    const fresh = await this.inner.fetch(feed);
    this.cache.set(key, { update: fresh, fetchedAtMs: Date.now() });
    return fresh;
  }

  /** Number of cached entries. */
  size(): number {
    return this.cache.size;
  }

  /** Drops every cached entry. */
  clear(): void {
    this.cache.clear();
  }
}

File: crates/pricefeed-client/src/cache.rs
//! A TTL cache decorator that wraps any [`PriceTransport`].

use std::collections::HashMap;
use std::time::{Duration, Instant};

use async_trait::async_trait;
use parking_lot::Mutex;
use pricefeed_types::{FeedId, PriceUpdate};

use crate::error::ClientError;
use crate::transport::PriceTransport;

struct Entry {
    update: PriceUpdate,
    fetched_at: Instant,
}

/// Wraps an inner transport with a time-to-live in-memory cache.
///
/// This is the decorator pattern: `CachingTransport<T>` *is a*
/// [`PriceTransport`], so it can be composed with the generic client exactly
/// like any other backend. Entries older than `ttl` are refreshed on access.
pub struct CachingTransport<T> {
    inner: T,
    ttl: Duration,
    cache: Mutex<HashMap<FeedId, Entry>>,
}

impl<T> CachingTransport<T> {
    /// Wraps `inner`, caching responses for `ttl`.
    #[must_use]
    pub fn new(inner: T, ttl: Duration) -> Self {
        Self {
            inner,
            ttl,
            cache: Mutex::new(HashMap::new()),
        }
    }

    /// The number of currently cached entries.
    #[must_use]
    pub fn len(&self) -> usize {
        self.cache.lock().len()
    }

    /// Whether the cache is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.cache.lock().is_empty()
    }

    /// Drops every cached entry.
    pub fn clear(&self) {
        self.cache.lock().clear();
    }

    fn cached_fresh(&self, feed: &FeedId) -> Option<PriceUpdate> {
        let cache = self.cache.lock();
        cache.get(feed).and_then(|e| {
            if e.fetched_at.elapsed() < self.ttl {
                Some(e.update.clone())
            } else {
                None
            }
        })
    }

    fn store(&self, feed: FeedId, update: PriceUpdate) {
        self.cache.lock().insert(
            feed,
            Entry {
                update,
                fetched_at: Instant::now(),
            },
        );
    }
}

#[async_trait]
impl<T: PriceTransport> PriceTransport for CachingTransport<T> {
    async fn fetch(&self, feed: &FeedId) -> Result<PriceUpdate, ClientError> {
        if let Some(hit) = self.cached_fresh(feed) {
            return Ok(hit);
        }
        let fresh = self.inner.fetch(feed).await?;
        self.store(feed.clone(), fresh.clone());
        Ok(fresh)
    }
}
